server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -267,8 +339,56 @@ private void downloadToFS(String fileName, Path location, String primaryTerm) th
             }
         }
 
-        // Mark in FileTransferTracker so that the same files are not uploaded at the time of translog sync
+        // Mark file download status as success in FileTransferTracker, this also gives us information to delete this file from remote
         fileTransferTracker.add(fileName, true);
+        // Mark in forTransferTracker that translog files for generation is downloaded from remote
+        fileTransferTracker.addGeneration(Long.parseLong(generation), true);
+    }
+
+    private void writeCkpFileFromMetadata(Map<String, String> metadata, Path location, String generation, String fileName)
+        throws IOException {
+
+        boolean downloadStatus = false;
+        long bytesToRead = 0;
+        try {
+            String ckpFileName = Translog.getCommitCheckpointFileName(Long.parseLong(generation));
+            Path filePath = location.resolve(ckpFileName);
+            // Here, we always override the existing file if present.
+            deleteFileIfExists(filePath);
+
+            String ckpDataBase64 = metadata.get(CHECKPOINT_FILE_DATA_KEY);
+            if (ckpDataBase64 == null) {
+                throw new IllegalStateException(
+                    "Checkpoint file data (key - ckp-data) is expected but not found in metadata for file: " + fileName
+                );
+            }
+            byte[] ckpFileBytes = Base64.getDecoder().decode(ckpDataBase64);
+            bytesToRead = ckpFileBytes.length;
+
+            Files.write(filePath, ckpFileBytes);
+            downloadStatus = true;
+
+            // Mark in FileTransferTracker that translog files for the generation is recovered
+            fileTransferTracker.addGeneration(Long.parseLong(generation), true);
+        } catch (IllegalStateException e) {
+            logger.error("Error processing metadata for translog file: {}", fileName);
+            throw e;
```

Comment:
Is this really required? can you not log in the if method itself before throwing the exception?